File: Python/serverselect.py
########################################################################################
# Author: Tianlong Song
# Name: serverselect.py
# Description: Socket server using select() to handle clients instead of threads
# Date created: 01/01/2015
########################################################################################

import socket,select

HOST = ""   # Symbolic name, meaning all available interfaces
PORT = 8888 # Arbitrary non-privileged port
CONNECTION_LIST = [] # List of sockets
DICT = {} # Use a dictionary to store active (socket,addr) pairs

server_socket = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
print "Server socket created"

# Bind socket to local host and port
try:
	server_socket.bind((HOST,PORT))
except socket.error as msg:
    	print "Bind failed. Error Code: " + str(msg[0]) + " Message " + msg[1]
    	sys.exit()
print "Socket bind complete"

# Start listening on socket
server_socket.listen(10)
print "Socket now listening"

# Add server socket to the list of readable connections
CONNECTION_LIST.append(server_socket);
print "Server started on port " + str(PORT)

# Talking with the client(s)
while True:
    	# Get the list sockets which are ready to be read through select
    	read_sockets,write_sockets,error_sockets = select.select(CONNECTION_LIST,[],[])

    	for sock in read_sockets:
        	# Deal with new connection
        	if sock==server_socket:
            		sockfd, addr = server_socket.accept()
            		CONNECTION_LIST.append(sockfd)
	    		DICT[sockfd] = addr
	    		print "Client (%s:%s) connected" % addr
	    		# Sending message to connected client
            		sockfd.send("Welcome to the server. Type something and hit enter")
        	# Some incoming message from a client
        	else:
            		try:
				data = sock.recv(1024)
				if data=="q" or data=="Q": 
					print "Client (%s:%s) is offline" % DICT[sock]
					sock.close()
					CONNECTION_LIST.remove(sock)
					del DICT[sock]
				elif data:
		    			reply = "You sent: " + data
		    			sock.sendall(reply)
            		except:
                		sock.close()
                		CONNECTION_LIST.remove(sock)
				del DICT[sock]

# Close the socket
server_socket.close()
print "Server socket closed"

File: C/serverselect.c
//**************************************************************************************
// Author: Tianlong Song
// Name: serverselect.c
// Description: Socket server using select to handle clinets
// Date created: 01/05/2015
//*************************************************************************************

#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#define PORT 8888	// The port number to be used
#define BACKLOG 10	// Maximum length of connection queue

int main() {
	struct sockaddr_in server,client;
	int sockfd,newfd;
	char message[1024];
	int c,i;
	const char *s = "Welcome to the server. Type something and hit enter.";
	
	fd_set master; // Master file descriptor list
	fd_set read_fds; // Temp file descriptor list
	int fdmax; // Maximum file descriptor number

	// Clear the master and temp list
	FD_ZERO(&master);
	FD_ZERO(&read_fds);

	// Settings
	server.sin_family = AF_INET; // IPV4
	server.sin_addr.s_addr = INADDR_ANY; // Any address
	server.sin_port = htons(PORT); //Port #, convert byte order to network format
	
	// Create server socket
	sockfd = socket(AF_INET,SOCK_STREAM,0);
	if(sockfd<0) {
		printf("Could not create socket!\n");
	}
	printf("Server socket created\n");

	// Bind socket to local host and port
	if(bind(sockfd,(struct sockaddr *)&server,sizeof(server))<0) {
		printf("Socket bind failed!\n");
	}
	printf("Socket bind complete\n");
	
	// Start listening on socket
	listen(sockfd,BACKLOG);

	// Add the server socket to the master set
	FD_SET(sockfd,&master);

	// Keep track of the biggest file descriptor number
	fdmax = sockfd;

	// Handling incoming connections
	while(1) {
		// Selection
		read_fds = master;
		if(select(fdmax+1,&read_fds,NULL,NULL,NULL)<0) {
			printf("Selection error!\n");
		}

		// Check each file descriptor
		for(i=0;i<=fdmax;i++) {
			if(FD_ISSET(i,&read_fds)) {
				if(i==sockfd) { // Server socket
					c = sizeof(struct sockaddr_in);
					newfd = accept(sockfd,(struct sockaddr *)&client,(socklen_t *)&c);
					if(newfd<0) {
						printf("Socket accept failed!\n");
					}
					printf("Connected to %s:%d\n",inet_ntoa(client.sin_addr),client.sin_port);
					memcpy(message,s,strlen(s));
					if(send(newfd,s,(int)strlen(s),0)+1<0) {
						printf("Send failed!\n");
					}
					// Add the new socket to master set
					FD_SET(newfd,&master);
					if(newfd>fdmax) { // Keep track of the max file descriptor
						fdmax = newfd;
					}
				} else { // Handle data from a client
					if(recv(i,message,1024,0)<0) {
						printf("Receive failed!\n");
					} else if((strcmp(message,"q")==0)||(strcmp(message,"Q")==0)) {
						printf("Client %s:%d is offline\n",inet_ntoa(client.sin_addr),client.sin_port);
						// Remove the quitting socket from master set
						FD_CLR(i,&master);
						if(fdmax==i) {
							fdmax--;
						}
					} else {
						if(send(i,message,(int)strlen(message)+1,0)<0) {
							printf("Send failed!\n");
						}
					}
				}
			}
		}
	}

	return 0;
}
